rmf_obstacle_ros2/src/lane_blocker/LaneBlocker.cpp

```diff
@@ -0,0 +1,964 @@
+void LaneBlocker::publish_speed_limit_req_msgs()
+{
+  for (auto& [_, msg] : _speed_limit_req_msgs)
+  {
+    if (msg->speed_limits.empty() && msg->remove_limits.empty())
+    {
+      RCLCPP_DEBUG(
+        this->get_logger(),
+        "None of the lanes for fleet %s have speed limits modified",
+        msg->fleet_name.c_str()
+      );
+      continue;
+    }
+    RCLCPP_INFO(
+      this->get_logger(),
+      "Requested %ld lanes to adhere to speed limit %f, for fleet %s",
+      msg->speed_limits.size(), _speed_limit, msg->fleet_name.c_str()
+    );
+    RCLCPP_INFO(
+      this->get_logger(),
+      "Requested %ld lanes to remove speed limit, for fleet %s",
+      msg->remove_limits.size(), msg->fleet_name.c_str()
+    );
+    _speed_limit_pub->publish(std::move(msg));
+  }
+}
+
+//==============================================================================
+std::optional<std::pair<std::string, std::size_t>> LaneBlocker::deserialize_key(
+  const std::string& key) const
+{
+  const std::string delimiter = "_";
+  try
+  {
+    std::string name = key.substr(0, key.find(delimiter));
+    std::string id_str =
+      key.substr(key.find(delimiter) + 1, key.size() - name.size());
+    std::stringstream ss(id_str);
+    std::size_t id; ss >> id;
+    return std::make_pair(name, id);
+  }
+  catch (const std::exception& e)
+  {
+    RCLCPP_INFO(
+      this->get_logger(),
+      "[LaneBlocker::deserialize_key] Unable to parse key. This is a bug and "
+      "should be reported. Detailed error: %s", e.what());
+    return std::nullopt;
+  }
+}
+
+//==============================================================================
+void LaneBlocker::purge_obstacles(
+  const std::unordered_set<std::string>& obstacle_keys,
+  const bool erase_from_buffer)
+{
+  for (const auto& obs : obstacle_keys)
+  {
+    auto lanes_it = _obstacle_to_lanes_map.find(obs);
+    if (lanes_it != _obstacle_to_lanes_map.end())
+    {
+      const auto& lanes = lanes_it->second;
+      for (const auto& lane_key : lanes)
+      {
+        auto obs_it = _lane_to_obstacles_map.find(lane_key);
+        if (obs_it != _lane_to_obstacles_map.end())
+        {
+          obs_it->second.erase(obs);
+        }
+      }
+    }
+    _obstacle_to_lanes_map.erase(obs);
+    if (erase_from_buffer)
+      _obstacle_buffer.erase(obs);
+  }
+}
+
+//==============================================================================
+void LaneBlocker::cull()
+{
+  // Cull obstacles that are past their expiry times.
+  // Also decide whether previously closed lanes should be re-opened.
+  std::unordered_set<std::string> obstacles_to_cull;
+  std::unordered_set<std::string> lanes_with_changes;
+
+  const auto now = this->get_clock()->now();
+  for (const auto& [obstacle_key, lanes] : _obstacle_to_lanes_map)
+  {
+    auto it = _obstacle_buffer.find(obstacle_key);
+    if (it == _obstacle_buffer.end())
+    {
+      // TODO(YV): Purge
```

Comment:
@Yadunund  is this TODO done ? I can see this method called `purge_obstacles(obstacles_to_cull);` ?